force-app/main/default/classes/SOQL.cls

```diff
@@ -270,12 +384,28 @@ public with sharing class SOQL {
      * @return List of contacts with the same Mailing State as the Account Billing State
      */
     public static List<Contact> getAccountsWithContactsByState(Id accountId) {
-        // Query the Account
-
-        // Store the Account's BillingState
-
-        // Query the Contacts
-        return null; // Replace null with the variable you used to store the result
+        List<Contact> finalList = new List<Contact>();
+        List<Account> accountList = new List<Account>([
+            SELECT Id, Name, BillingState, (SELECT Id, MailingState FROM Contacts)
+            FROM Account
+            WHERE Id =:accountId WITH SECURITY_ENFORCED
+        ]);
+        
+        if (accountList[0] != null) {
+            String billingState = accountList[0].BillingState;
+        List<Contact> contactList = new List<Contact>([
+            SELECT Id, LastName, AccountId, MailingState
+            FROM Contact 
+            WHERE MailingState =:billingState WITH SECURITY_ENFORCED
+        ]);
+
+        for (Contact con : contactList) {
```

Comment:
You shouldn't need this for loop to find contacts with MailingState = billingState because you are already filtering for this directly in your Contact query